lualib/resty/events/broker.lua

```diff
@@ -28,64 +25,125 @@ local decode = codec.decode
-            str_sub(err, -11) == "broken pipe")
+local function terminating(self, worker_connection)
+    return self._clients[worker_connection] == nil or exiting()
 end
 
--- broadcast to all/unique workers
-local function broadcast_events(self, unique, data)
-    local n = 0
-
-    -- if unique, schedule to a random worker
-    local idx = unique and random(1, nkeys(self._clients))
+local function get_event_data(self, event_data)
+    local unique = event_data.spec.unique
+    if unique then
+        local uniques = self._uniques
+        if uniques:get(unique) then
+            if not exiting() then
+                log(DEBUG, "unique event is duplicate: ", unique)
+            end
 
-    for _, q in pairs(self._clients) do
+            return
+        end
 
-        -- skip some and broadcast to one workers
-        if unique then
-            idx = idx - 1
+        uniques:set(unique, 1, self._opts.unique_timeout)
+    end
+    return event_data.data, unique
+end
 
-            if idx > 0 then
-                goto continue
-            end
-        end
+-- broadcast to all/unique workers
+local function broadcast_events(self, event_data)
+    local data, unique = get_event_data(self, event_data)
+    if not data then
+        return
+    end
 
-        local ok, err = q:push(data)
+    local n = 0
 
-        if not ok then
+    -- pairs is "random" enough for unique
+    for _, client_queue in pairs(self._clients) do
+        local _, err = client_queue:push(data)
+        if err then
             log(ERR, "failed to publish event: ", err, ". ",
                      "data is :", cjson_encode(decode(data)))
-
         else
             n = n + 1
 
             if unique then
                 break
             end
         end
+    end
+
+    log(DEBUG, "event published to ", n, " workers")
+end
+
+local function read_thread(self, worker_connection)
+    while not terminating(self, worker_connection) do
+        local data, err = worker_connection:recv_frame()
+        if err then
+            if not is_timeout(err) then
+                return nil, "failed to read event from worker: " .. err
+            end
+
+            -- timeout
+            goto continue
+        end
+
+        if not data then
+            if not exiting() then
+                log(ERR, "did not receive event from worker")
+            end
+            goto continue
+        end
+
+        local event_data, err = decode(data)
+        if not event_data then
+            if not exiting() then
+                log(ERR, "failed to decode event data: ", err)
+            end
+            goto continue
+        end
+
+        broadcast_events(self, event_data)
 
         ::continue::
-    end  -- for q in pairs(_clients)
+    end -- while not terminating
 
-    log(DEBUG, "event published to ", n, " workers")
+    return true
+end
+
+local function write_thread(self, worker_connection)
+    while not terminating(self, worker_connection) do
```

Comment:
Should we add a line before `while`:
```
local q = self._clients[worker_connection]
```
because `self._clients[worker_connection]` may be `nil` when error occurs (line 177)